File: 백준/Platinum/1786. 찾기/찾기.py
import sys

T = sys.stdin.readline().rstrip()
P = sys.stdin.readline().rstrip()
lps = [0]*len(P)
ans = []

# 접두사와 접미사가 일치하는 가장 긴 길이 저장
def makeLps(p, lps):
    length = 0

    i = 1
    while i < len(p):
        if p[length] == p[i]:
            length += 1
            lps[i] = length
            i += 1

        else:
            if length == 0:
                lps[i] = 0
                i += 1
            else:
                length = lps[length - 1]



def KMP(T, P):
    lenT, lenP = len(T), len(P)
    i, j = 0, 0
    while i < lenT:
        # 동일하면 전진
        if T[i]==P[j]:
            i+=1
            j+=1
        # 동일하지 않고 보고있는게 0번째가 아니면
        # 봐야하는 lps부터 보기
        elif j!=0:
            j = lps[j-1]
        # 동일하지 않고 0번째면 그냥 다음으로
        else:
            i+=1

        # print(i, j)
        if j==lenP:
            ans.append(i-lenP+1)
            j = lps[j-1]

makeLps(P, lps)
KMP(T,P)

print(len(ans))
if ans:
    print(*ans)

File: 백준/Platinum/1786. 찾기/찾기.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<int,int> pii;
#define FOR(i,a,b) for(int i=(a);i<=(b);i++)
#define MAX

string s,t;
int lps[1000010];
vector<int> v;

// 접두사와 접미사가 일치하는 가장 긴 길이 저장
void makeLps(){
    int len=0;
    for(int i=1;i<t.length();i++){
        while(len>0 && t[i]!=t[len]){
            len=lps[len-1];
        }
        if(t[i]==t[len]) lps[i]=++len;
    }
}

void kmp() {
    int j=0;
    for(int i=0;i<s.length();i++){
        while(j>0 && s[i]!=t[j])
            j = lps[j-1];
        if(s[i] == t[j]){
            if(j==t.length()-1){
                // cout << i << '\n';
                v.push_back(i-t.length()+1);
                j = lps[j];
            }
            else j++;
        }
    }
}

int main(){
    ios_base::sync_with_stdio(0); cin.tie(0);
    
    getline(cin,s);
    getline(cin,t);
    makeLps();
    kmp();

    cout<<v.size()<<'\n';

    for (auto i:v) {
        cout << i+1 << ' ';
    }
    return 0;
}